consensus/src/quorum_store/proof_manager.rs

```diff
@@ -203,92 +204,116 @@ impl ProofManager {
-                    if proof_block.is_empty() && inline_block.is_empty() {
-                        Payload::empty(true, self.allow_batches_without_pos_in_proposal)
-                    } else if inline_block.is_empty() {
-                        trace!(
-                            "QS: GetBlockRequest excluded len {}, block len {}",
-                            excluded_batches.len(),
-                            proof_block.len()
+                let cur_txns = txns_with_proof_size + opt_batch_txns_size;
+                let (inline_block, inline_block_size) =
+                    if self.allow_batches_without_pos_in_proposal && proof_queue_fully_utilized {
+                        let mut max_inline_txns_to_pull = min(
+                            request.max_txns.saturating_sub(cur_txns),
+                            request.max_inline_txns,
                         );
-                        Payload::InQuorumStore(ProofWithData::new(proof_block))
-                    } else {
-                        trace!(
-                            "QS: GetBlockRequest excluded len {}, block len {}, inline len {}",
-                            excluded_batches.len(),
-                            proof_block.len(),
-                            inline_block.len()
+                        max_inline_txns_to_pull.count = min(
+                            max_inline_txns_to_pull.count,
+                            request.max_unique_txns.saturating_sub(cur_unique_txns),
                         );
-                        Payload::QuorumStoreInlineHybrid(
-                            inline_block,
-                            ProofWithData::new(proof_block),
-                            None,
+                        self.batch_queue.pull_batches(
+                            max_inline_txns_to_pull,
+                            excluded_batches
+                                .iter()
+                                .cloned()
+                                .chain(proof_block.iter().map(|proof| proof.info().clone()))
+                                .collect(),
                         )
-                    },
-                );
-                match callback.send(Ok(res)) {
+                    } else {
+                        (Vec::new(), PayloadTxnsSize::zero())
+                    };
+                counters::NUM_INLINE_BATCHES.observe(inline_block_size.count as f64);
+                counters::NUM_INLINE_TXNS.observe(inline_block_size.bytes as f64);
+
+                let response = if self.enable_opt_quorum_store {
+                    Payload::OptQuorumStore(OptQuorumStorePayload::new(
+                        inline_block
+                            .into_iter()
+                            .map(|(info, txns)| InlineBatch::new(info, txns))
+                            .collect::<Vec<InlineBatch>>()
+                            .into(),
+                        opt_batches.into(),
+                        proof_block.into(),
```

Comment:
I think you are missing one part of the packing.
Currently, if we do not have opt_batches and inline txns we will have half-empty blocks.
I think you should try to add more proofs at the end (check if there is more space available after pulling opt_batches and inline and used it for proofs).

Reply:
Although it is very unlikely to have more proofs than batches. Some Byzantine behavior probably has to be involved in this case.

Reply:
I will address this with heuristic changes.